crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -61,30 +67,24 @@ impl Scoring {
 // fees). Also, executed amount contains the fees for sell order.
 #[derive(Debug, Clone)]
 pub struct Trade {
-    sell: eth::Asset,
-    buy: eth::Asset,
-    side: Side,
+    fulfillment: Fulfillment,
     executed: order::TargetAmount,
+    uniform_prices: ClearingPrices,
     custom_price: CustomClearingPrices,
-    policies: Vec<order::FeePolicy>,
 }
 
 impl Trade {
     pub fn new(
-        sell: eth::Asset,
-        buy: eth::Asset,
-        side: Side,
+        fulfillment: &Fulfillment,
         executed: order::TargetAmount,
+        uniform_prices: ClearingPrices,
```

Comment:
Let's check if it is possible to calculate the same thing without using the `Fulfillment` and uniform clearing prices. This is very important to achieve because it proves that the same calculation can be done on an onchain observable settlement (where `fulfillment.executed` and `fulfillment.fee` are not visible and also where uniform clearing prices also potentially won't be visible soon).

So, if my math is correct, we should be able to:
1. Calculate the protocol fee based on the `Trade` data as it was before this PR
2. Create new Trade with the same data as old Trade but with adjusted custom prices to exclude protocol fee =>  new_custom_prices
3. Calculate next protocol fee using newly created Trade

How do we get from original custom prices to new custom prices while working only with protocol fee:

```
let custom_price = match trade.order().side {
                order::Side::Sell => scoring::CustomClearingPrices {
                    sell: custom_price.sell + protocol_fee,
                    buy: custom_price.buy,
                },
                order::Side::Buy => scoring::CustomClearingPrices {
                    sell: custom_price.sell,
                    buy: custom_price.buy - protocol_fee,
                },
            };
```
(don't ask me how I came to this) 😄 The math is just a bit more complex if driver doesn't use the original trade values to build the custom prices (for example if scales them), but we can tackle that later.

Can we try this out and see if it works?

Reply:
Thanks @sunce86 , it works amazingly!!